File: 알고리즘/leetcode/seoin/array/TwoSum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        d = {}
        for i, j in enumerate(nums):
            r = target - j
            if r in d: return [d[r], i]
            d[j] = i

        # for i in range(len(nums)):
        #     for j in range(i + 1, len(nums)):
        #         if (nums[i] + nums[j] == target):
        #             return [i, j]

File: 알고리즘/leetcode/seoin/array/TwoSum.kt
class Solution {
    fun twoSum(nums: IntArray, target: Int): IntArray {
        val lens = nums.size
        val ans : IntArray = intArrayOf(-1, -1)
        for (i in 0 until lens) {
            for (j in i + 1 until lens) {
                if (nums[i] + nums[j] == target)  {
                    ans[0] = i
                    ans[1] = j
                    break
                }
            }
            if (ans[0] != -1) break
        }
        return ans
    }
}

class Solution {
    fun twoSum(nums: IntArray, target: Int): IntArray {
        val indexMap = mutableMapOf<Int, Int>()
        for (i in nums.indices) {
            val desired = target - nums[i]
            indexMap[desired]?.let {
                return intArrayOf(it, i)
            }
        indexMap[nums[i]] = i
        }

    throw IllegalStateException("The problem should have at least one solution")
    }
}

